gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -424,6 +418,24 @@ private GoogleCloudStorageImpl(
     this.downscopedAccessTokenFn = downscopedAccessTokenFn;
   }
 
+  private static MetricsRecorder getMetricsRecorder(
```

Comment:
IIUC, What we are trying to do here is to enforce the use the `credential` which was passed in to constructor instead of extracting it from `RetryHttpInitializer`.
In production code flow we ALWAYS pass in the `credentials` which is used in creating the HTTP transport for `Storage`. So, `credential` is/can never `null` in production code flow. I am up for adding a null check for that.

Now, coming to testing, we agree that we need multiple GCSImpl constructors so that we can mock or override the storage OR transport layer. In any case we must have used some credentials (either a fake or real), why can't we update our test case to make sure to use the constructor where `credential` is getting passed instead of just passing in the overriden transport layer? I do agree that it will require a bit more changes in our tests but IMO I will still prefer that instead of having hacks in production code where God knows which credential is getting used for what.
Another point in favour of not adding this hack would be that GCSImpl is going to be here long enough (even after integrating new java-client) if possible let's try not to add hack in here yet.

On high level I believe extracting `credential` from http transport is wrong (And maybe extracting Http transport form storage too). It will never fall under production code flow and for testing it's `test` code responsibility to pass in the correct credential.
Will let @medb @cnauroth Comment on this.